dune_client/api/extensions.py

```diff
@@ -0,0 +1,196 @@
+        Executes a Dune query, waits till execution completes,
+        fetches and the results in CSV format
+        (use it load the data directly in pandas.from_csv() or similar frameworks)
+        """
+        job_id = self._refresh(
+            query, ping_frequency=ping_frequency, performance=performance
+        )
+        return self.get_execution_results_csv(job_id)
+
+    def run_query_dataframe(
+        self, query: QueryBase, performance: Optional[str] = None
+    ) -> Any:
+        """
+        Execute a Dune Query, waits till execution completes,
+        fetched and returns the result as a Pandas DataFrame
+
+        This is a convenience method that uses run_query_csv() + pandas.read_csv() underneath
+        """
+        try:
+            import pandas  # type: ignore # pylint: disable=import-outside-toplevel
+        except ImportError as exc:
+            raise ImportError(
+                "dependency failure, pandas is required but missing"
+            ) from exc
+        data = self.run_query_csv(query, performance=performance).data
+        return pandas.read_csv(data)
+
+    def get_latest_result(self, query: Union[QueryBase, str, int]) -> ResultsResponse:
+        """
+        GET the latest results for a query_id without re-executing the query
+        (doesn't use execution credits)
+
+        :param query: :class:`Query` object OR query id as string | int
+
+        https://dune.com/docs/api/api-reference/latest_results/
+        """
+        if isinstance(query, QueryBase):
+            params = {
+                f"params.{p.key}": p.to_dict()["value"] for p in query.parameters()
+            }
+            query_id = query.query_id
+        else:
+            params = None
+            query_id = int(query)
+
+        response_json = self._get(
+            route=f"/query/{query_id}/results",
+            params=params,
+        )
+        try:
+            return ResultsResponse.from_dict(response_json)
+        except KeyError as err:
+            raise DuneError(response_json, "ResultsResponse", err) from err
+
+    def upload_csv(self, table_name: str, data: str, description: str = "") -> bool:
+        """
+        https://dune.com/docs/api/api-reference/upload-data/?h=data+upload#endpoint
+        The write API allows you to upload any .csv file into Dune. The only limitations are:
+
+        - File has to be < 200 MB
+        - Column names in the table can't start with a special character or digits.
+
+        Below are the specifics of how to work with the API.
+        """
+        response_json = self._post(
+            route="/table/upload/csv",
+            params={
+                "table_name": table_name,
+                "description": description,
+                "data": data,
+            },
+        )
+        try:
+            return bool(response_json["success"])
+        except KeyError as err:
+            raise DuneError(response_json, "upload_csv response", err) from err
```

Comment:
I am guessing you are referring to the fact that this is boolean and not a fancy class. Then this winds up being less consistent than the rest or the types. It just seemed a bit overkill to create a whole class just for a boolean value.

Reply:
Sorry, I could have made my comment clearer; it is in fact much smaller and nitpicky.

What I was getting at is for instance, for cancelling an execution, you also don't create a class because it's a boolean: https://github.com/duneanalytics/dune-client/blob/c4cfb38a0041bbd7651bf7d86a9aead03d968530/dune_client/api/execution.py#L54-L59

But in the exception, you use `CancellationResponse` but here, it is formatted differently (`upload_csv response`).

Reply:
OH! I see now. Right, if that's what you mean, then I will just pass through and at least make them more uniform. Do you have a preference for style? `CamelCaseResponse` or `function_name`? I think I'll go with the camel case.